### PR TITLE
Kernel/aarch64: Actually make the MemoryManager work!

### PR DESCRIPTION
**Pull request by FireFox317**
Previously we were able to compile `MemoryManager`, however it wasn't actually possible to do anything useful with it. This PR implements the quickmap mechanism and fixes the missing bits to actually map memory and create Regions etc.

For now the aarch64 Kernel is identity mapped, but later on we could change this to actually use the same loading addresses as the x86 kernel.

image

### CODE REVIEW

**Comment by nico** [comment]
typos in commit messages:
- `tlb flusing` => `tlb flushing`
- `Setup foo` => `Set up foo`

File: Kernel/Memory/MemoryManager.cpp [hunk 1]
```diff
@@ -243,7 +243,9 @@ UNMAP_AFTER_INIT void MemoryManager::parse_memory_map()
     // Register used memory regions that we know of.
     m_global_data.with([&](auto& global_data) {
         global_data.used_memory_ranges.ensure_capacity(4);
+#if ARCH(I386) || ARCH(X86_64)
         global_data.used_memory_ranges.append(UsedMemoryRange { UsedMemoryRangeType::LowMemory, PhysicalAddress(0x00000000), PhysicalAddress(1 * MiB) });
```
Comment (nico) [inline Kernel/Memory/MemoryManager.cpp:253]: Is it not necessary to map the zero page, to make null pointer derefs fault?

**Comment by nico** [comment]
Does this have any effect on boot speed?

File: Kernel/Memory/MemoryManager.cpp (hunk 1 above)
Comment (FireFox317) [reply Kernel/Memory/MemoryManager.cpp:253]: The `used_memory_ranges` are physical memory ranges that are already occupied in physical memory. For example for `x86` the 0-1Mib physical range is used for BIOS related things IIRC. However the kernel is actually running in the virtual memory space, where the zero page currently is unmapped, thus causing a fault when it is accessed.

**Comment by FireFox317** [comment]
Thanks for the review! I have updated the commit messages.

I haven't measured whether it actually boots faster, but I think it is still a good idea to enable the MMU as soon as possible in the boot process, because we can enable caches at that point. For example the kmalloc_init actually memsets it's memory to zero, which will definitely faster with caches enabled.

**Review by nico — APPROVED** [review]
Looks good to me, FWIW.

File: Kernel/Arch/aarch64/init.cpp [hunk 2]
```diff
@@ -113,6 +115,9 @@ extern "C" [[noreturn]] void init()
     new (&bootstrap_processor()) Processor();
```
Comment (bugreport0) [inline Kernel/Arch/aarch64/init.cpp:115]: In `init.cpp:106` should we update 'Serenity OS' to 'SerenityOS' for consistency? I can't add a comment at that exact line, so sneaking it in here.

Comment (nico) [reply Kernel/Arch/aarch64/init.cpp:115]: Seems unrelatedly to this PR. Feel free to make a PR for that (but that string is temporary anyways) 🙂

Comment (bugreport0) [reply Kernel/Arch/aarch64/init.cpp:115]: Okay, sorry for the noise!

File: Kernel/Arch/PageDirectory.h [hunk 3]
```diff
@@ -1,156 +1,17 @@
 /*
- * Copyright (c) 2018-2021, Andreas Kling <kling@serenityos.org>
+ * Copyright (c) 2022, Timon Kruiper <timonkruiper@gmail.com>
```
Comment (bgianfo) [inline Kernel/Arch/PageDirectory.h:2]: These are normally only additive, it's (IMHO) in bad taste to remove someone else's copyright like.

Comment (FireFox317) [reply Kernel/Arch/PageDirectory.h:3]: Yeah true, the file basically completely changed, but anyway it's indeed better to keep the copyright. I pushed changes for this!

**Pull request merged**